src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java

```
@@ -0,0 +1,112 @@
+                    return f;
+                }
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public String[] getCommand() {
+        String[] decorated = cmds;
+        if (shellDecorator != null) {
+            decorated = shellDecorator.decorate(decorated);
+        }
+        return decorated;
+    }
+
+    @Override
+    protected Process createProcess() {
+        WindowsRunasProcess winProcess = new WindowsRunasProcess(null, userDecorator.getUser());
+        winProcess.setLpEnvironment(computeEnvironmentBlock());
+        winProcess.setLpCurrentDirectory(dir.getAbsolutePath());
+        winProcess.start(String.join(" ", cmds));
+        return winProcess;
+    }
+
+    /**
+     * Convert environment Map to lpEnvironment block format.
+     * @return environment block for starting a process
+     */
+    private String computeEnvironmentBlock() {
+        // TODO: copy over the source code so that it's the same regardless of Java version this is running on
+        // http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/windows/classes/java/lang/ProcessEnvironment.java
+        String block = null;
+        try {
+            Class<?> cl = Class.forName("java.lang.ProcessEnvironment");
+            for (final Method method : cl.getDeclaredMethods()) {
+                boolean correctMethod =
+                        "toEnvironmentBlock".equals(method.getName()) && method.getParameterTypes().length == 1
+                                && method.getParameterTypes()[0].equals(Map.class);
+                if (correctMethod) {
+                    method.setAccessible(true);
+                    block = (String) method.invoke(null, environment);
+                    break;
+                }
+            }
+        } catch (Throwable t) {
+            logger.atError().setCause(t)
+                    .log("Failed to compute environment block. Creating process without env modifications");
+        }
```

Is it okay to create process fail silently? E.g. if it's for a lifecycle script the only way for users to find out about failures will be logs

---

Good point. This should fail the process creation. Things won't work properly without env vars